uniswap_edpr/src/logics/shared_algorithm_lens/src/lib.rs

```diff
@@ -0,0 +1,86 @@
+  let pool_fees_result = get_get_last_snapshot_value_in_pool_fees (targets . get (0usize) . unwrap () . clone ()) . await ;
+  let tc28x6_result = get_get_last_snapshot_value_in_tcumul_28x6hr (targets . get (1usize) . unwrap () . clone ()) . await ;
+  let v3pool_result = get_get_last_snapshot_value_in_v3pool (targets . get (2usize) . unwrap () . clone ()) . await ;
+  let eth_usdc_price_result = get_get_last_snapshot_value_in_eth_usdc_price (targets . get (3usize) . unwrap () . clone ()) . await ;
+  
+  let fees_24h_usd = pool_fees_result.clone().unwrap().result[0].fees_24h_usd;
+  let t0_decimals = pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t0_decimals;
+  let t1_decimals =  pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t1_decimals;
+
+  let eth_price_usd  = eth_usdc_price_result.clone().unwrap().result.t1_price_usd;
+
+  let address = v3pool_result.clone().unwrap().result.address;
+  let current_tick_liquidity_str = v3pool_result.clone().unwrap().result.liquidity;
+  let current_tick_liquidity = i128::from_str_radix(current_tick_liquidity_str.trim_start_matches("0x"), 16).unwrap();
+  let sqrt_ratio_x96_str = v3pool_result.clone().unwrap().result.sqrt_ratio_x96;
+  let sqrt_ratio_x96 = u128::from_str_radix(sqrt_ratio_x96_str.trim_start_matches("0x"), 16).unwrap();
+  let tick_current = v3pool_result.clone().unwrap().result.tick_current;
+  let tick_spacing = v3pool_result.clone().unwrap().result.tick_spacing;
+  let token0 = v3pool_result.clone().unwrap().result.token0;
+
+  let tick_cumul_28x6h = tc28x6_result.unwrap().0;
+
+  let current_sqrt_price = sqrt_ratio_x96 as f32 / f32::powf(2.0,96.0);
+  let mut current_price = current_sqrt_price * current_sqrt_price;
+
+  let mut compressed: i32 = tick_current / tick_spacing;
+  if tick_current <0 && tick_current % tick_spacing != 0 {
+    compressed -= 1;
+  }
+  let floor = compressed * tick_spacing;
+  let mut min_tick = floor;
+  let mut max_tick = floor + tick_spacing;
+  for i in (1..29).rev() {
+    let new_tick = ((tick_cumul_28x6h[i] - tick_cumul_28x6h[i-1]) / 21600) as i32;
+    if new_tick < min_tick {
+      min_tick = new_tick;
+    } else if new_tick > max_tick {
+      max_tick = new_tick;
+    }
+  }
+  
+  let mut range_top = f32::powf(1.0001, max_tick as f32);
+  let mut range_bottom = f32::powf(1.0001, min_tick as f32);
+  let amount0 = current_tick_liquidity as f32
+    * (range_top.sqrt() - current_sqrt_price as f32)
+    / (current_sqrt_price as f32 * range_top.sqrt())
+    / f32::powf(10.0, t0_decimals as f32);
+  let amount1 = current_tick_liquidity as f32
+    * (current_sqrt_price as f32 - range_bottom.sqrt())
+    / f32::powf(10.0, t1_decimals as f32);
+  let mut est_tvl_in_range = amount0 * current_price + amount1;
+
+  if token0 == "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2" {
+    current_price = 1.0 / current_price;
+    est_tvl_in_range = amount0 + amount1 * current_price;
```

Comment:
I understand that you want to convert to ETH based values.
Is it not this way?
```suggestion
    est_tvl_in_range = amount0 + amount1 / current_price;
```

Reply:
https://github.com/horizonx-tech/chainsight-showcase/pull/34#issuecomment-1890470697

Reply:
> current_price in this part of the code represents ETH per RPL rather than the other way around

Really? I see sqrt_ratio_x96 as being close to the sqrtPriceX96 in the following,
https://blog.uniswap.org/uniswap-v3-math-primer
Isn't this value inverted to match token0, token1 in that order?
ex:
DAI/ETH -> token0=DAI,token1=ETH, current_price = DAI per ETH
ETH/DAI -> token0=ETH,token1=DAI, current_price = Same as for DAI/ETH?

Reply:
sqrt_ratio_x96 is the same as sqrtPriceX96. The notation in the blog is a bit confusing and ambiguous. There's one particular paragraph in the blog where I would eliminate the middle of the paragraph to avoid the confusion and just read it as

"In the USDC-WETH 5-bps pool (also the .05% fee tier), token0 for this pool is USDC and token 1 is WETH. ... For this specific pool, this exchange rate [i.e. the price ratio] is the amount of WETH that could be traded for 1 USDC."

Our code works and gives the correct result as it is, so I would just leave it there.

Reply:
I finally figured it out: if token0 = WETH, it doesn't matter because you invert the current_price before the code you are talking about in the branch when token0 = WETH. Sorry it took so long.
But this is the downside of mutable variables.  As with range_top and range_bottom, the inverted current_price should be assigned to another variable for use.

```rust
```

Reply:
No problem.

This part is necessary because the variables are being swapped around as well as inverted:
```
    let new_range_top = 1.0 / range_bottom;
    let new_range_bottom = 1.0 / range_top;
    range_top = new_range_top;
    range_bottom = new_range_bottom;
```

The other part I've written is potentially confusing as you've noted, but just needs careful reading. To my mind the other part you've written seems unnecessary and less efficient, but it's still reasonably straightforward so I've gone ahead and made this change.

Reply:
This should be improved as it will reduce the possibility of misreading. I certainly supported a more verbose writing style. I would like to see an end to the use of variables that are inherently mutable. (Vectors, for example, that make multiple updates may be used, but you are just replacing them according to token0)
Your comment is also correct and should be modified to stop using mutable variables.

I think the method will be similar to the comments here.
https://github.com/horizonx-tech/chainsight-showcase/pull/34#discussion_r1450529551

Reply:
modified to clean.
https://github.com/horizonx-tech/chainsight-showcase/pull/34/commits/2a37216167d7eb76ab16e230b94729bd094e4b10